src/ledger/database/lmdb.zig

```diff
@@ -0,0 +1,716 @@
+    key: []const u8,
+    operation: CursorAbsoluteOperation,
+) LmdbError!?struct { []const u8, []const u8 } {
+    var key_val = toVal(key);
+    var val_val: c.MDB_val = undefined;
+    result(c.mdb_cursor_get(
+        cursor,
+        &key_val,
+        &val_val,
+        @intFromEnum(operation),
+    )) catch |err| switch (err) {
+        error.MDB_NOTFOUND => return null,
+        else => return err,
+    };
+    return .{ fromVal(key_val), fromVal(val_val) };
+}
+
+fn cursorGetRelative(
+    cursor: *c.MDB_cursor,
+    operation: CursorRelativeOperation,
+) LmdbError!?struct { []const u8, []const u8 } {
+    var key_val: c.MDB_val = undefined;
+    var val_val: c.MDB_val = undefined;
+    result(c.mdb_cursor_get(
+        cursor,
+        &key_val,
+        &val_val,
+        @intFromEnum(operation),
+    )) catch |err| switch (err) {
+        error.MDB_NOTFOUND => return null,
+        else => return err,
+    };
+    return .{ fromVal(key_val), fromVal(val_val) };
+}
+
+/// Cursor Get operations that require a key to execute a lookup
+const CursorAbsoluteOperation = enum(c_uint) {
+    /// Position at key/data pair. Only for #MDB_DUPSORT
+    get_both = 2,
+    /// position at key, nearest data. Only for #MDB_DUPSORT
+    get_both_range = 3,
+
+    /// Position at specified key
+    set = 15,
+    /// Position at specified key, return key + data
+    set_key = 16,
+    /// Position at first key greater than or equal to specified key.
+    set_range = 17,
+};
+
+/// Cursor Get operations that do *not* require a key to execute a lookup
+const CursorRelativeOperation = enum(c_uint) {
+    /// Position at first key/data item
+    first = 0,
+    /// Position at first data item of current key. Only for #MDB_DUPSORT
+    first_dup = 1,
+
+    /// Return key/data at current cursor position
+    get_current = 4,
+    /// Return up to a page of duplicate data items from current cursor position.
+    /// Move cursor to prepare for #MDB_NEXT_MULTIPLE. Only for #MDB_DUPFIXED
+    get_multiple = 5,
+    /// Position at last key/data item
+    last = 6,
+    /// Position at last data item of current key. Only for #MDB_DUPSORT
+    last_dup = 7,
+    /// Position at next data item
+    next = 8,
+    /// Position at next data item of current key. Only for #MDB_DUPSORT
+    next_dup = 9,
+    /// Return up to a page of duplicate data items from next cursor position.
+    /// Move cursor to prepare for #MDB_NEXT_MULTIPLE. Only for #MDB_DUPFIXED
+    next_multiple = 10,
+    /// Position at first data item of next key
+    next_nodup = 11,
+    /// Position at previous data item
+    prev = 12,
+    /// Position at previous data item of current key. Only for #MDB_DUPSORT
+    prev_dup = 13,
+    /// Position at last data item of previous key
+    prev_nodup = 14,
+
+    /// Position at previous page and return up to a page of duplicate data items.
+    /// Only for #MDB_DUPFIXED
+    prev_multiple = 18,
+};
+
+fn result(int: isize) LmdbError!void {
```

Comment:
I don't think `result` is the best name for this. 

These seems more like a "convert errno to error set" function, so I'd recommend it returns an error set and then be wrapped in a "check for error return code" function if that makes sense.

Reply:
I'd love to find a better name for this, but I don't see the point of splitting it into two functions.

Assuming the name is descriptive, I feel that it helps readability for the name to be concise, since it's used for every call to lmdb. I'm struggling to find a name that feels satisfying in that regard. I think "lmdbReturnCodeToErrorUnion" is sufficiently descriptive, but it feels like clutter.

Reply:
How about `maybeError`, which reads nicely in `try maybeError(int)`, ie "try unwrap this which is maybe an error".

Reply:
6f3c9ccf6846de4a4638886851f56daf092a9b1f